cub/warp/warp_exchange.cuh

```diff
@@ -468,5 +468,89 @@ public:
   //@}  end member group
 };
 
+template <typename InputT, int ITEMS_PER_THREAD>
+class WarpExchangeRegister
+{
+  static_assert(PowerOfTwo<ITEMS_PER_THREAD>::VALUE, "ITEMS_PER_THREAD must be a power of two");
+
+  template <int SIZE, int idx>
+  class CompileTimeArrayImpl; // fwd decl for partial specializations
+
+  template <int SIZE> // terminating partial specialization
+  class CompileTimeArrayImpl<SIZE, SIZE>
+  {
+  protected:
+    static_assert(PowerOfTwo<SIZE>::VALUE, "SIZE must be a power of two");
+
+    // used for dumping back the individual values after transposing
+    InputT (&output_items)[SIZE];
+
+    template <int>
+    __device__ void Loop(bool, unsigned)
+    {}
+
+  public:
+    __device__ CompileTimeArrayImpl(InputT (&input_items)[SIZE])
+        : output_items{input_items}
+    {}
+  };
+
+  template <int SIZE, int idx> // concrete recursion class
+  class CompileTimeArrayImpl : protected CompileTimeArrayImpl<SIZE, idx + 1>
+  {
+  protected:
+    static_assert(PowerOfTwo<SIZE>::VALUE, "SIZE must be a power of two");
+
+    InputT val;
+
+    template <int NUM_ENTRIES>
+    __device__ void Loop(const bool xor_bit_set, const unsigned mask)
+    {
+      // NOTE: Do *NOT* try to refactor this code to use a reference, since nvcc
+      //       tends to choke on it and then drop everything into local memory.
+      const InputT send_val = (xor_bit_set ? CompileTimeArrayImpl<SIZE, idx>::val
+                                           : CompileTimeArrayImpl<SIZE, idx + NUM_ENTRIES>::val);
+      const InputT recv_val = __shfl_xor_sync(mask, send_val, NUM_ENTRIES, SIZE);
+      (xor_bit_set ? CompileTimeArrayImpl<SIZE, idx>::val
+                   : CompileTimeArrayImpl<SIZE, idx + NUM_ENTRIES>::val) = recv_val;
+
+      constexpr int next_idx = idx + 1 + ((idx + 1) % NUM_ENTRIES == 0) * NUM_ENTRIES;
+      CompileTimeArrayImpl<SIZE, next_idx>::template Loop<NUM_ENTRIES>(xor_bit_set, mask);
+    }
+
+    // terminate recursion
+    template <int NUM_ENTRIES, std::enable_if_t<!NUM_ENTRIES, int> = 0>
+    __device__ void TransposeImpl(const unsigned mask)
+    {}
+
+    template <int NUM_ENTRIES, std::enable_if_t<NUM_ENTRIES, int> = 0>
+    __device__ void TransposeImpl(const unsigned mask)
+    {
+      const bool xor_bit_set = threadIdx.x & NUM_ENTRIES;
+      Loop<NUM_ENTRIES>(xor_bit_set, mask);
+
+      TransposeImpl<NUM_ENTRIES / 2>(mask);
+    }
+
+  public:
+    __device__ CompileTimeArrayImpl(InputT (&input_items)[SIZE])
+        : CompileTimeArrayImpl<SIZE, idx + 1>{input_items}
+        , val{input_items[idx]}
+    {}
+
+    __device__ ~CompileTimeArrayImpl() { this->output_items[idx] = val; }
+
+    __device__ void Transpose(const unsigned mask) { TransposeImpl<SIZE / 2>(mask); }
+  };
+
+  using CompileTimeArray = CompileTimeArrayImpl<ITEMS_PER_THREAD, 0>;
+
+public:
+  __device__ void Transpose(InputT (&input_items)[ITEMS_PER_THREAD], const unsigned mask = 0xFFFF'FFFFU) const {
```

Comment:
Our warp-scope facilities usually support multiple logical warps when you partition the actual warp into smaller contiguous groups of threads. If you refactor the code to be an implementation of `cub::WarpExchange`, you might need to accept lane id and the mask and use those instead of `threadIdx.x`.

Reply:
sure, is there a portable way to get the `lane_id` in form of `0 <= lane_id < ITEMS_PER_THREAD`?

Reply:
`WarpExchange` already have `lane_id` and `mask` computed. Your struct can receive these from `WarpExchange`.